### PR TITLE
Clarify offset rules a bit

### PR DESCRIPTION
**Pull request by nagisa**

### CODE REVIEW

**Comment by nagisa** [comment]
cc @Gankro

**Comment by Gankro** [comment]
This matches my understanding, but I'd like someone more... qualified to chime in. See: http://llvm.org/docs/LangRef.html#id195

In particular it's not clear to me that undefined behaviour is actually invoked if the result is unused. LLVM says it just produces a [poison value](http://llvm.org/docs/LangRef.html#poison-values).

File: src/libcore/intrinsics.rs [hunk 1]
```diff
@@ -255,12 +255,16 @@ extern "rust-intrinsic" {
...
+    /// Calculates the offset from a pointer.
     ///
     /// This is implemented as an intrinsic to avoid converting to and from an
     /// integer, since the conversion would throw away aliasing information.
+    ///
+    /// # Safety
+    ///
+    /// Both the starting and resulting pointer must be either in bounds or one byte past the end
+    /// of an allocated object. If either pointer is out of bounds or arithmetic overflow occurs,
+    /// undefined behaviour is invoked, regardless of whether the pointer is used or not.
```
Comment (alexcrichton) [inline src/libcore/intrinsics.rs]: Could you keep these comments wrapped at 80 chars? (surrounding style)

**Comment by alexcrichton** [comment]
I agree with @Gankro that I think the language here may be a bit strong, as generating `undef` or poison values I believe is not undefined behavior, it's just later using them which may cause undefined behavior. I'd need clarification on that, however, as well.

**Comment by alexcrichton** [comment]
@nagisa could you clarify these docs to something along the lines of:

> # Safety
> 
> Both the starting and resulting pointer must be either in bounds or one byte past the end of an allocated object. If either pointer is out of bounds or arithmetic overflow occurs then any further use of the returned value will result in undefined behavior.

LLVM's docs about poison values [explicitly says](http://llvm.org/docs/LangRef.html#poisonvalues):

> Poison values have the same behavior as undef values, with the additional effect that any instruction that has a dependence on a poison value has undefined behavior.

An `undef` (to my knowledge) is just a bag-of-bits, so it's undefined behavior if you load from an address whose value is `undef` (for example), but arithmetic/addition on the value is defined (it's just weird).

**Comment by nagisa** [comment]
Done.

**Comment by alexcrichton** [comment]
@bors: r+ 82b43568a64822c693fdfa50a736056efe92c4ce rollup

Thanks!

**Pull request merged**